### PR TITLE
systemd: Block login until config stage completes

### PR DESCRIPTION
**Pull request by holmanb**
```
systemd: Block login until config stage completes

LP: #2013403
```

## Additional Context
Discovered on riscv64, cc_set_passwords runs in cloud config stage, but tty access blocked by cloud init stage. This means that a user may be able to try logging in prior to the user account being configured.

## Test Steps

download the ubuntu riscv unmatched image, install riscv qemu dependencies, then: 
```
qemu-system-riscv64 \
    -machine virt \
    -cpu rv64 \
    -m 1G \
    -device virtio-blk-device,drive=hd \
    -drive file=lunar-preinstalled-server-riscv64+unmatched.img,if=none,id=hd \
    -device virtio-net-device,netdev=net \
    -netdev user,id=net,hostfwd=tcp::2222-:22 \
    -bios /usr/lib/riscv64-linux-gnu/opensbi/generic/fw_jump.elf \
    -kernel /usr/lib/u-boot/qemu-riscv64_smode/uboot.elf \
    -object rng-random,filename=/dev/urandom,id=rng \
    -device virtio-rng-device,rng=rng \
    -append "root=LABEL=rootfs console=ttyS0" \
    -nographic
```

The prompt becomes available prior to cloud-init configuring the user.

### CODE REVIEW

**Review by j5awry — APPROVED** [review]
community +1 from the Canonical Public Cloud team!

**Comment by TheRealFalcon** [comment]
Given that we can't ssh until the user is created, I'm not foreseeing any huge differences in boot speed on the standard cloud boot to ssh path, but it'd be good to double check. I think it's probably worth asking Paride if he can run the boot speed jobs against this branch to see if there's any noticeable difference.

**Comment by holmanb** [comment]
See the before/after critical-chain logs below. Note that this change blocks `systemd-user-sessions.service` until `cloud-config.service` is complete, which delays the prompt until the user has been configured, as required.

### Before
multi-user.target
```bash
ubuntu@ubuntu:~$ systemd-analyze critical-chain multi-user.target
The time when unit became active or started is printed after the "@" character.
The time the unit took to start is printed after the "+" character.

multi-user.target @2min 8.647s
└─snapd.seeded.service @1min 46.006s +3.547s
  └─snapd.service @1min 21.051s +24.895s
    └─basic.target @1min 20.744s
      └─sockets.target @1min 20.741s
        └─snapd.socket @1min 20.658s +76ms
          └─sysinit.target @1min 20.483s
            └─cloud-init.service @57.819s +22.636s
              └─systemd-networkd-wait-online.service @57.407s +372ms
                └─systemd-networkd.service @56.165s +1.186s
                  └─network-pre.target @56.124s
                    └─cloud-init-local.service @17.107s +39.005s
                      └─systemd-remount-fs.service @16.275s +785ms
                        └─systemd-fsck-root.service @13.215s +3.013s
                          └─systemd-journald.socket @12.218s
                            └─-.mount @12.016s
                              └─-.slice @12.015s

```

systemd-user-sessions:
```bash
ubuntu@ubuntu:~$ systemd-analyze critical-chain systemd-user-sessions.service
The time when unit became active or started is printed after the "@" character.
The time the unit took to start is printed after the "+" character.

systemd-user-sessions.service +745ms
└─remote-fs.target @1min 23.065s
  └─remote-fs-pre.target @1min 23.059s
```

### After
multi-user.target
```
ubuntu@ubuntu:~$ systemd-analyze critical-chain multi-user.target
The time when unit became active or started is printed after the "@" character.
The time the unit took to start is printed after the "+" character.

multi-user.target @2min 6.845s
└─getty.target @2min 6.836s
  └─getty@tty1.service @2min 6.814s
    └─system-getty.slice @2min 6.766s
      └─setvtrgb.service @2min 6.588s +138ms
        └─plymouth-quit-wait.service @2min 6.291s +167ms
          └─systemd-user-sessions.service @2min 6.096s +140ms
            └─cloud-config.service @1min 47.007s +19.018s
              └─snapd.seeded.service @1min 43.410s +3.539s
                └─snapd.service @1min 18.880s +24.414s
                  └─basic.target @1min 18.587s
                    └─sockets.target @1min 18.583s
                      └─snapd.socket @1min 18.512s +65ms
                        └─sysinit.target @1min 18.357s
                          └─cloud-init.service @57.836s +20.502s
                            └─systemd-networkd-wait-online.service @57.461s +32>
                              └─systemd-networkd.service @56.114s +1.295s
                                └─network-pre.target @56.071s
                                  └─cloud-init-local.service @14.584s +41.473s
                                    └─systemd-remount-fs.service @13.599s +939ms
```

systemd-user-sessions:
```bash
ubuntu@ubuntu:~$ systemd-analyze critical-chain systemd-user-sessions.service
The time when unit became active or started is printed after the "@" character.
The time the unit took to start is printed after the "+" character.

systemd-user-sessions.service +140ms
└─cloud-config.service @1min 47.007s +19.018s
  └─snapd.seeded.service @1min 43.410s +3.539s
    └─snapd.service @1min 18.880s +24.414s
      └─basic.target @1min 18.587s
        └─sockets.target @1min 18.583s
          └─snapd.socket @1min 18.512s +65ms
            └─sysinit.target @1min 18.357s
              └─cloud-init.service @57.836s +20.502s
                └─systemd-networkd-wait-online.service @57.461s +325ms
                  └─systemd-networkd.service @56.114s +1.295s
                    └─network-pre.target @56.071s
                      └─cloud-init-local.service @14.584s +41.473s
                        └─systemd-remount-fs.service @13.599s +939ms
                          └─systemd-fsck-root.service @10.795s +2.759s
                            └─systemd-journald.socket @9.850s
                              └─-.mount @9.611s
                                └─-.slice @9.608s

**Comment by TheRealFalcon** [comment]
@paride Do we have the ability to run bootspeed tests here?

**Comment by paride** [comment]
TL;DR: I see no performance regression, I think this is good to land.

---

Here are some benchmark result done as follows:

- Tesdbed machine: "bare" qemu/kvm VM (no LXD or libvirt)
- Architecture: amd64
- Ubuntu release under test: Jammy
- No dhcp (direct forward of port 22)

First I prepared an image with cloud-init built from the tip of the main branch (5674a7dc), performed measurement, then rebased the `holmanb/block-tty` branch on main, rebuilt/reinstalled the package and measured again. The "base" image I used is https://cloud-images.ubuntu.com/jammy/20230616/jammy-server-cloudimg-amd64-disk-kvm.img. After installing the custom package I performed a `cloud-init clean` to make cloud-init re-run. Each runs has simple cloud-init very basic user-data from a localds datasource:
```
#cloud-config
ssh_authorized_keys:
  - ssh-rsa ...
```

Instances were started with:

```
qemu-system-x86_64  \
  -pidfile pidfile \
  -machine accel=kvm,type=q35 \
  -cpu host \
  -m 1G \
  -nographic \
  -display none \
  -device virtio-net-pci,netdev=net0 \
  -netdev user,id=net0,hostfwd=tcp::2222-:22 \
  -drive if=virtio,format=qcow2,file=jammy-server-cloudimg-amd64-sut.img \
  -drive if=virtio,format=raw,file=seed.img
```

The `benchmark.sh` script basically contains a loop like this:
```
while true; do
    ssh -o ConnectTimeout=1 -o LogLevel=ERROR -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no -p 2222 ubuntu@127.0.0.1 cloud-init status --wait && break
done
```

Measurements were performed with `hyperfine`. Each measurement was performed 20 times.

#### cloud-init from main, time to first ssh login

| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `./benchmark.sh` | 16.103 ± 1.119 | 14.055 | 17.403 | 1.00 |

Median: 16.59s

#### cloud-init from main, time to `cloud-init status --wait`

| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `./benchmark.sh` | 21.193 ± 0.827 | 19.706 | 22.152 | 1.00 |

Median: 21.52s

#### cloud-init from block-tty, time to first ssh login

| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `./benchmark.sh` | 16.166 ± 1.037 | 13.908 | 17.337 | 1.00 |

Median: 16.51s.

#### cloud-init from block-tty, time to `cloud-init status --wait`

| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `./benchmark.sh` | 21.186 ± 0.864 | 19.629 | 22.362 | 1.00 |

Median: 21.39s

**Review by paride — APPROVED** [review]
Rationale, diff and benchmark results LGTM.

**Comment by paride** [comment]
(It would be nice to have some kind of performance regression test like this in CI, but we woulds need a bare metal runner to have a "quiet" environment and allow virtualization.)

**Comment by blackboxsw** [comment]
@holmanb I think we can proceed with this approach then

**Review by blackboxsw — APPROVED** [review]
LGTM, negligible bootspeed impact and/or time to SSH for this shift.
